### ISSUE TITLE
Feature Request: Thread Initialization and Termination hooks

### ISSUE THREAD

**Issue by snikopoulos** [grpc/grpc#18158]
### What version of gRPC and what language are you using?

gRPC v1.7.3 and now v1.17.2
C++
 
### What operating system (Linux, Windows,...) and version?

Linux - Centos7x
Windows Server 2012R2 +
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-36)
Visual Studio 2017

 
### Problem statement

We are using gRPC to provide service for the [Domino](https://www.ibm.com/collaboration/ibm-domino) server. The Domino runtime requires that threads using the C API have been initialized and when they exit that they also call a thread terminate function to release allocated resources.

gRPC service worker threads need to be initialized and terminated so they can use the Domino runtime. gRPC provides hooks that are executed before and after every request. This option does not work well, because the init/term overhead is high compared to the cost of executing the actual service request.

We need a thread initialization/termination hook for all service worker threads that is executed once at the start of a thread and once at the end of the thread. The cost of init/term is then taken only once and is outside the scope of executing a service request.

### What did you expect to see?
 
We have an implementation of this feature for gRPC v1.7.3 and we're moving it up to the most current release (v1.17).

Would like to know if this change (or something similar) can be merged into your project. I can provide a pull request if you believe you might be interested.

The high points of the changes are:

1) new method in ServerBuilder class

```
  /// Register ThreadCallbacks
  ServerBuilder& RegisterThreadCallbacks(grpc::Server::ThreadCallbacks *thread_callbacks) {
      thread_callbacks_ = thread_callbacks;
      return *this;
  }
```

2) new grpc::Server::ThreadCallbacks class

```
  class ThreadCallbacks {
   public:
     virtual void ThreadInit() = 0;
     virtual void ThreadTerm() = 0;
  };
```

3) callouts added to the WorkerThread::Run method 

```
 void ThreadManager::WorkerThread::Run() {
   thd_mgr_->thread_callbacks_->ThreadInit();
   thd_mgr_->MainWorkLoop();
   thd_mgr_->MarkAsCompleted(this);
   thd_mgr_->thread_callbacks_->ThreadTerm();
 }
```

**Comment by yang-g** [comment]
I do not think this would be something used commonly.
I think there are two ways potentially to achieve what you want.
The first one is to override gRPC core thread creation in the similar way of thd_posix vs thd_windows by defining GPR_CUSTOM_SYNC.
The second one is for grpc::ServerBuilder to accept a passed in thread pool to handle sync requests. (which we do not have right now).
